src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs

```diff
@@ -0,0 +1,373 @@
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Host;
+using Microsoft.Azure.WebJobs.Script.Binding;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Scripting;
+using Microsoft.CodeAnalysis.Scripting;
+
+namespace Microsoft.Azure.WebJobs.Script.Description
+{
+    public class CSharpFunctionInvoker : ScriptFunctionInvokerBase
+    {
+        private const string ScriptClassName = "Submission#0";
+
+        private readonly ScriptHost _host;
+        private readonly DictionaryJsonConverter _dictionaryJsonConverter = new DictionaryJsonConverter();
+        private readonly TraceWriter _fileTraceWriter;
+        private readonly FunctionMetadata _functionMetadata;
+        private readonly BindingMetadata _trigger;
+        private readonly Collection<FunctionBinding> _inputBindings;
+        private readonly Collection<FunctionBinding> _outputBindings;
+        private readonly FileSystemWatcher _fileWatcher;
+        private readonly string _triggerParameterName;
+        private readonly bool _omitInputParameter;
+        private readonly IFunctionEntryPointResolver _functionEntryPointResolver;
+        private MethodInfo _function;
+
+        public CSharpFunctionInvoker(ScriptHost host, BindingMetadata trigger, FunctionMetadata metadata, bool omitInputParameter,
+            Collection<FunctionBinding> inputBindings, Collection<FunctionBinding> outputBindings, IFunctionEntryPointResolver functionEntryPointResolver)
+        {
+            _host = host;
+            _trigger = trigger;
+            _triggerParameterName = trigger.Name;
+            _omitInputParameter = omitInputParameter;
+            _inputBindings = inputBindings;
+            _outputBindings = outputBindings;
+            _functionMetadata = metadata;
+            _functionEntryPointResolver = functionEntryPointResolver;
+            _fileWatcher = InitializeFileWatcher(host);
+
+            if (_host.ScriptConfig.FileLoggingEnabled)
+            {
+                string logFilePath = Path.Combine(_host.ScriptConfig.RootLogPath, "Function", _functionMetadata.Name);
+                _fileTraceWriter = new FileTraceWriter(logFilePath, TraceLevel.Verbose);
+            }
+            else
+            {
+                _fileTraceWriter = NullTraceWriter.Instance;
+            }
+        }
+
+        public static string[] DefaultAssemblyReferences
+            => new string[]
+            {
+                "System",
+                "System.Core",
+                "System.Xml",
+                "System.Net.Http",
+                "Microsoft.WindowsAzure.Storage"
+            };
+
+        public static Assembly[] DefaultAssemblies
+            => new Assembly[]
+            {
+                typeof(TraceWriter).Assembly,
+                typeof(object).Assembly,
+                typeof(TimerInfo).Assembly
+            };
```

Comment:
Yes, we'll be using Project.json and the functionality provided to us by NuGet to install/restore packages and generate the lock file we'll be able to use to grab the references we'll need to pass into Roslyn (the last bit _may_ require some custom code, but we might be able to leverage the work other teams are doing, we're investigating), so we won't be reinventing the wheel with nugetsm, and external references work with the approach we're following as well.